docs/proposals/20210112-encryption-roviders.md

```diff
@@ -0,0 +1,95 @@
+# Encryption Providers for encrypted secrets at rest
+
+**Auther**: Mohamed Elsayed (@moelsayed)
+**Status**: Draft
+
+
+## Abstract
+
+By default, all Kubernetes secret objects are stored on disk in plain text inside etcd. The [Encryption Providers](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/
+) feature was added to Kubernets starting with version 1.13. 
+
+At rest data encryption is a requirement for security compliance and adds an additional layer of security for secret data, especially when etcd nodes are separated from the control plan and in off-node backups. 
+
+KubeOne needs to support this feature natively. Meaning the user should be able to enable, disable the feature and rotate keys when needed without having to apply any actions manually. 
+
+## Goals
+
+* Provide a safe path to enable/disable Encryption Providers.
+* Support atomic(?) rotation for existing keys.
+* Rewriting all secret resources (no just secrets) after enable/disable/rotate operations.
+
+## Non-Goals
+
+* Support external KMS providers.
+
+## Challenges
+
+The feature has a lot of moving parts; as it requires performing a specific sequence of actions, including changing the KubeAPI configuration, restarting KubeAPI and rewriting all secret resources to apply the encryption. This requires the implementation to be as idempotent as possible with ability to rollback on failure, with out breaking the cluster. 
+
+## Implementation
+
+Unfortunately, it's not possible to simply update the KubeAPI configuration and expect the configuration to reconcile. KubeOne will have to _read_ the _current_ configuration on the cluster, _mutate_ it based on the _required_ state and then apply it. Additionally, KubeOne will have to be able to revert changes on any errors and recover safely if the process is interrupted at any point.
+
+The configuration for this will be added under `features` in the KubeOneCluster spec:
+
+```yaml
+apiVersion: kubeone.io/v1beta1
+kind: KubeOneCluster
+features:
+  encryptionProviders:
+    enabled: true
```

Comment:
Keeping in mind future API extensions that will be inevitably required, simple boolean flag is not enough. Flags are needed per provider type, and not only flags but possible config options too.

So we are looking at 3 possible providers:
* none
* static (the one for being implemented now)
* kms (the one for being implemented later)

Those should be expressed as additional objects inside `encryptionProviders`, like

```
#### VERY DRAFT, MUCH FANTASY ####
####### only for demo purpose #######
apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
features:
  encryptionProviders:
    - kind: none
    - kind: static
    - kind: kms
      endpoint: unix:///tmp/socketfile.sock
      cachesize: 100
      timeout: 3s
```

Alternatively, we could not reimplement parts of what's already done in upstream and "just" provide the

```
apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
features:
  encryptionConfiguration: |
    apiVersion: apiserver.config.k8s.io/v1
    kind: EncryptionConfiguration
    resources:
    - resources:
      - secrets
      providers:
      - kms:
          name: myKmsPlugin
          endpoint: unix:///tmp/socketfile.sock
          cachesize: 100
          timeout: 3s
      - identity: {}
```

Given that second option, we can actually extract the `endpoint` from KMS providers and automatically mount them to the kube-apiserver pod ( with extra mounts in kubeadm config).

Provided that we can detect changes in `encryptionConfiguration` (compare actual file to encryptionConfiguration`, we can automate and orchestrate kubeapi pod restarts.

Reply:
Providing ONLY `encryptionConfiguration` we could give all encryption options to the end user, including KMS support in one shot.

Reply:
I agree with what @kron4eg said, but in that case, I'd leave the `enable` field in the API and implement it like @moelsayed initially proposed. For example:

```yaml
features:
  encryptionProviders:
    # enable/disable 
    enable: true
    # setting this value will disable automated rotation.
    customProviderFile: |
      apiVersion: apiserver.config.k8s.io/v1
      kind: EncryptionConfiguration
      ...
```

The difference between the originally proposed API and this API is that I've removed the `rotate` field. Instead of the `rotate` field, we would keep the CLI flag as it's currently proposed.

* If `enable` is set to `false`
  * Do nothing if the feature has never been enabled
  * Otherwise, disable the feature
* If `enable` is set to `true`:
  * If `customProviderFile` is not set, generate it ourselves and handle key management
  * If `customProviderFile` is set, upload it to the instance. Additionally, if KMS is configured, create the appropriate volumes on the API server pods
  * Configure the API server and everything else as described in the proposal

Note: if `customProviderFile` is set and the CLI flag is set, KubeOne would either show a warning or fail.

In this case, we can quickly implement support for all setups. Technically, users could use Terraform/Ansible to deploy KMS and with such API, they could actually use it in their clusters. I think that it wouldn't require too much additional effort on our side, but we could solve everything (besides deploying KMS, which should be discussed) in a single run.

Reply:
@kron4eg @xmudrii I updated the proposal accordingly. Please check.

For the initial implementation, I will stop at implementing the the custom config basic functionality. I will create a new issue/PR to address deploying the KMS and creating the required volume mounts for the API server. I am considering that out of scope for the initial implementation.

Reply:
@moelsayed Can you create an Epic for this?

Reply:
@xmudrii there is already an epic to track the proposal and the implementation: https://github.com/kubermatic/kubeone/issues/769

Reply:
@moelsayed I was thinking about an Epic for addressing the KMS deployment and the volume creation. You can also extend the current Epic. I only see issues relevant to the initial implementation in the Epic, but please correct me if I'm wrong.

Reply:
What happens if:
* `enable: true`, but no `customProviderFile` is given (e.g. it's empty)?
* `enable: true`, but no `customProviderFile` is given and install/apply is run, and then `customProviderFile` becomes defined?

Reply:
> enable: true, but no customProviderFile is given (e.g. it's empty)?
We manage everything automatically. A new config file is created and pushed to the nodes. In this case key rotation is supported. 
> enable: true, but no customProviderFile is given and install/apply is run, and then customProviderFile becomes defined?
K1 should push the defined custom file but it will refuse to the rotate the keys if the rotate flag is set.

Reply:
@xmudrii I updated the epic with an issue to add KMS #1242.